jwst/master_background/master_background_step.py

```diff
@@ -0,0 +1,86 @@
+from ..stpipe import Step
+from .. import datamodels
+
+
+__all__ = ["MasterBackgroundStep"]
+
+
+class MasterBackgroundStep(Step):
+    """
+    MasterBackgroundStep:  Compute and subtract master background from spectra
+    """
+
+    spec = """
+        user_background = str(default=None) # Path to user-supplied master background
+        save_background = boolean(default=False) # Save computed master background
+    """
+
+
+    def process(self, input):
+
+        """
+        Compute a master background from N > 1 dedicated background exposures
+        and subtract it from the target science exposure(s)
+
+        Parameters
+        ----------
+        input : `~jwst.datamodels.ImageModel`, `~jwst.datamodels.IFUImageModel`, `~jwst.datamodels.ModelContainer`, association
+            Input target data model(s) to which master background subtraction is
+            to be applied
+
+        user_background: string path or `~jwst.datamodels.MultiSpecModel`
+            User-supplied master background 1D spectrum, path to file or opened
+            datamodel
+
+        Returns
+        -------
+        result: `~jwst.datamodels.ImageModel`, `~jwst.datamodels.IFUImageModel`, `~jwst.datamodels.ModelContainer`
+            The background-subtracted target data model(s)
+        """
+
+        with datamodels.open(input) as input_data:
+
+            # Handle individual NIRSpec FS, NIRSpec MOS
+            if isinstance(input_data, datamodels.MultiSlitModel):
+                pass
+
+            # Handle associations, or input ModelContainers
+            elif isinstance(input_data, datamodels.ModelContainer):
+                pass
+
+            # Handle MIRI LRS
+            elif isinstance(input_data, datamodels.ImageModel):
+                pass
+
+            # Handle MIRI MRS and NIRSpec IFU
+            elif isinstance(input_data, datamodels.IFUImageModel):
```

Comment:
Could IFUImageModel be handled the same as ImageModel?

Reply:
I think an `IFUImageModel` will have to have each slice iterated over?  But my understanding is that the slices are still embedded in the 2D `ImageModel`, right?  Maybe @jemorrison can clarify?

Reply:
Yes, an `IFUImageModel` is a 2D IFU image.

Reply:
There are a few differences in the content of `ImageModel` vs. `IFUImageModel`. For example, `ImageModel` includes things like the pathloss correction arrays, but `IFUImageModel` does not (because IFU data don't get that correction). So while the master background subtraction step may operate on both of them in the same way, it's still important to carry through the original input model type to the output that's created, because the output is intended to be an updated copy of the input (and hence we don't want to accidentally lose some extra data arrays along the way).

Reply:
The main difference is that the `IFUImageModel` has ~30 cutouts and WCS objects that need to be iterated over, whereas the MIRI LRS `ImageModel` has 1.  Is the way these are accessed different as well?  I.e. one has to tell the WCS object which slice to grab for the `IFUImageModel`?

Reply:
The IFU modes do not go through extract_2d, so no cutouts. Unless you meant something else.

Reply:
For NIRSpec there is a list of wcs objects, one for each IFU stripe.  Each stripe has a bounding_box that specifies where that stripe is on the full 2-D image.  I don't know how MIRI handles this.

Reply:
There is code in the NIRSpec portion of the flat-field step, as well as the NIRSpec IFU and MIRI MRS portions of the photom step, that already do the operation of building a 2-D wavelength "image" for those types of exposures, because the corrections/calibrations done in those steps are wavelength-dependent. So those can be used as examples. And don't we also carry along a 2-D wavelength array in the data products for spectroscopic modes now, or is that only for slit-like modes and not IFU? If the 2-D wavelength array is already in the product/model, you're all set.